pkg/multicast/multicast.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicast
+
+import (
+	"net"
+
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"golang.org/x/net/ipv6"
+)
+
+var (
+	// v6Socket is the udp socket used to join/leave a multicast group
+	v6Socket net.PacketConn
+
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "multicast")
+)
+
+// Pre-Defined Multicast Addresses
+// as defined in https://tools.ietf.org/html/rfc4291#section-2.7.1
+
+var (
+	// AllNodesIfcLocalMaddr is the multicast address that identifies the group of
+	// all IPv6 nodes, within scope 1 (interface-local)
+	AllNodesIfcLocalMaddr net.IP = []byte{
+		0xff, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1,
+	}
+
+	// AllNodesLinkLocalMaddr is the multicast address that identifies the group of
+	// all IPv6 nodes, within scope 2 (link-local)
+	AllNodesLinkLocalMaddr net.IP = []byte{
+		0xff, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1,
+	}
+
+	// AllRoutersIfcLocalMaddr is the multicast address that identifies the group of
+	// all IPv6 routers, within scope 1 (interface-local)
+	AllRoutersIfcLocalMaddr net.IP = []byte{
+		0xff, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x2,
+	}
+
+	// AllRoutersLinkLocalMaddr is the multicast address that identifies the group of
+	// all IPv6 routers, within scope 2 (link-local)
+	AllRoutersLinkLocalMaddr net.IP = []byte{
+		0xff, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x2,
+	}
+
+	// AllRoutersSiteLocalMaddr is the multicast address that identifies the group of
+	// all IPv6 routers, within scope 5 (site-local)
+	AllRoutersSiteLocalMaddr net.IP = []byte{
+		0xff, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x2,
+	}
+
+	// SolicitedNodeMaddrPrefix is the prefix of the multicast address that is used
+	// as part of NDP
+	SolicitedNodeMaddrPrefix net.IP = []byte{
+		0xff, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x1, 0xff, 0x0, 0x0, 0x0,
+	}
+)
+
+func init() {
+	c, err := net.ListenPacket("udp6", "[::]:22334")
+	if err != nil {
+		log.Warnf("Failed to listen on socket for multicast: %v", err)
+		log.Warnf("Multicast group join/leave functionality is disabled")
```

Comment:
Given the user explicitly asked for this (disabled by default), maybe we should log at least an error instead of a warning?

Reply:
It's in `init()` which executed regardless. I'll move it inside the functions and protect it by `atomic.Once`, so it only runs when needed.